File: ruby/lib/pdf_oxide/types/search_options.rb
# frozen_string_literal: true

module PdfOxide
  module Types
    # Options for text search operations
    class SearchOptions
      attr_accessor :case_sensitive, :whole_words_only, :regex, :include_annotations, :ignore_accents

      # Initialize search options
      # @param case_sensitive [Boolean] Case sensitive search
      # @param whole_words_only [Boolean] Match whole words only
      # @param regex [Boolean] Use regex pattern
      # @param include_annotations [Boolean] Include annotations in search
      # @param ignore_accents [Boolean] Ignore accents in search
      def initialize(
        case_sensitive: false,
        whole_words_only: false,
        regex: false,
        include_annotations: true,
        ignore_accents: false
      )
        @case_sensitive = case_sensitive
        @whole_words_only = whole_words_only
        @regex = regex
        @include_annotations = include_annotations
        @ignore_accents = ignore_accents
      end

      # Create case-insensitive search options
      # @return [SearchOptions] New instance
      def self.case_insensitive
        new(case_sensitive: false)
      end

      # Create case-sensitive search options
      # @return [SearchOptions] New instance
      def self.case_sensitive
        new(case_sensitive: true)
      end

      # Create regex search options
      # @return [SearchOptions] New instance
      def self.regex
        new(regex: true)
      end

      # Convert to hash
      # @return [Hash] Hash representation
      def to_h
        {
          case_sensitive: @case_sensitive,
          whole_words_only: @whole_words_only,
          regex: @regex,
          include_annotations: @include_annotations,
          ignore_accents: @ignore_accents
        }
      end

      # Convert to string
      # @return [String] String representation
      def to_s
        flags = []
        flags << 'case-sensitive' if @case_sensitive
        flags << 'whole-words' if @whole_words_only
        flags << 'regex' if @regex
        "SearchOptions(#{flags.join(', ')})"
      end
    end
  end
end

File: php/src/Builders/SearchOptions.php
<?php

declare(strict_types=1);

namespace PdfOxide\Builders;

/**
 * Options for PDF search operations.
 *
 * Provides fluent interface for configuring search behavior.
 */
class SearchOptions
{
    private bool $caseSensitive = false;
    private bool $wholeWordsOnly = false;
    private bool $useRegex = false;
    private int $maxResults = 0; // 0 = no limit
    private bool $includeAnnotations = false;
    private bool $searchHiddenText = false;
    private array $pageRange = []; // Empty = all pages

    /**
     * Set case sensitivity for search.
     */
    public function caseSensitive(bool $sensitive): self
    {
        $this->caseSensitive = $sensitive;
        return $this;
    }

    /**
     * Set whether to match whole words only.
     */
    public function wholeWordsOnly(bool $whole): self
    {
        $this->wholeWordsOnly = $whole;
        return $this;
    }

    /**
     * Set whether to use regex patterns.
     */
    public function useRegex(bool $regex): self
    {
        $this->useRegex = $regex;
        return $this;
    }

    /**
     * Set maximum number of results.
     */
    public function maxResults(int $max): self
    {
        $this->maxResults = max(0, $max);
        return $this;
    }

    /**
     * Set whether to search annotations.
     */
    public function includeAnnotations(bool $include): self
    {
        $this->includeAnnotations = $include;
        return $this;
    }

    /**
     * Set whether to search hidden text.
     */
    public function searchHiddenText(bool $search): self
    {
        $this->searchHiddenText = $search;
        return $this;
    }

    /**
     * Set page range for search.
     *
     * @param int $startPage Zero-based start page
     * @param int $endPage Zero-based end page (inclusive)
     */
    public function pageRange(int $startPage, int $endPage): self
    {
        $this->pageRange = [
            'start' => max(0, $startPage),
            'end' => max(0, $endPage),
        ];
        return $this;
    }

    // Getters
    public function isCaseSensitive(): bool { return $this->caseSensitive; }
    public function isWholeWordsOnly(): bool { return $this->wholeWordsOnly; }
    public function isUsingRegex(): bool { return $this->useRegex; }
    public function getMaxResults(): int { return $this->maxResults; }
    public function isIncludingAnnotations(): bool { return $this->includeAnnotations; }
    public function isSearchingHiddenText(): bool { return $this->searchHiddenText; }
    public function getPageRange(): array { return $this->pageRange; }

    /**
     * Convert to array for FFI calls.
     */
    public function toArray(): array
    {
        return [
            'case_sensitive' => $this->caseSensitive,
            'whole_words_only' => $this->wholeWordsOnly,
            'use_regex' => $this->useRegex,
            'max_results' => $this->maxResults,
            'include_annotations' => $this->includeAnnotations,
            'search_hidden_text' => $this->searchHiddenText,
            'page_range' => $this->pageRange,
        ];
    }

    /**
     * Create from array.
     */
    public static function fromArray(array $options): self
    {
        $opts = new self();

        if (isset($options['case_sensitive'])) {
            $opts->caseSensitive($options['case_sensitive']);
        }
        if (isset($options['whole_words_only'])) {
            $opts->wholeWordsOnly($options['whole_words_only']);
        }
        if (isset($options['use_regex'])) {
            $opts->useRegex($options['use_regex']);
        }
        if (isset($options['max_results'])) {
            $opts->maxResults($options['max_results']);
        }
        if (isset($options['include_annotations'])) {
            $opts->includeAnnotations($options['include_annotations']);
        }
        if (isset($options['search_hidden_text'])) {
            $opts->searchHiddenText($options['search_hidden_text']);
        }
        if (isset($options['page_range'])) {
            $opts->pageRange(
                $options['page_range']['start'] ?? 0,
                $options['page_range']['end'] ?? 0
            );
        }

        return $opts;
    }
}
